# Tracking Antimicrobial Resistance Organisms Timely (TAROT) for methicillin-resistance *Staphylococcus aureus* (TAROT-MRSA)

![Tracking Antimicrobial Resistant Organisms Timely_v3_trimed](https://github.com/user-attachments/assets/91b94402-258c-43ce-a836-eae9e4ce1fee)


# About TAROT
TAROT is a concept for performing successive and on-demand core-genome single nucleotide polymorphism (cgSNP)-based phylogenetic analysis using whole-genome sequencing of bacterial strains. It aims to support the improvement of infection prevention and control intervention in hospitals. We have developed and validated a bioinformatics pipeline for executing TAROT analyses.

# Intended Use
Sequence read data generated by Oxford Nanopore Technologies (ONT) sequencers can be used with this pipeline to analyze MRSA strains.

# Requirements
You need to use a computer with an x86 or x64 CPU architecture. Using a computer with ARM CPU architecture may cause issues when pulling the Docker image of the tarot-mrsa-env.  


# Preparering Environment Using Docker
You can set up an environment to run the TAROT-MRSA pipeline on your computer using Docker.

After installing Docker, pull the Docker image "tarot-mrsa-env" from Dockerhub using the following command: 
```docker image pull kotaroaoki/tarot-mrsa-env:latest```.

After that, run the container with this command; ```docker run --rm -it kotaroaoki/tarot-mrsa-env:latest bash```.

You can mount a host directory directory in the Docker container using the "-v" option. 

# Run TAROT-MRSA
1. Place MRSA strain sequence data generated using ONT v14 reagent and Flow Cell R10.4.1 into each strain ID directory under the "seq_data" directory.
2. Do change the directory containing the sample you want to analyze.
3. Run the script with the following command; ```bash /tmp/program/TAROT_MRSA_successive_cgSNP_analysis.sh```
6. Repeat from 2 to 3.

If you have analyzed three or more strains of the same sequence type you are analyzing, the script will perform a cgSNP analysis.  
